File: amsr/tokens.py
from re import compile, escape

# 12 dihedral symbols: 30 degree discretization
DIHEDRALS = ["^^", "^\\", "<\\", ">>", "</", "_/", "__", "\\_", "\\>", "<<", "/>", "/^"]

# 24 dihedral symbols: 15 degree discretization
# DIHEDRALS = [f"/{chr(x)}" for x in range(ord("a"), ord("y"))]

N_DIHEDRALS = len(DIHEDRALS)
Z = DIHEDRALS[0]
E = DIHEDRALS[N_DIHEDRALS // 2]
CW = "'"
CCW = "`"
PLUS = "+"
MINUS = "-"
EXTRA_PI = ":"
BANG = "!"
RADICAL = "*"
L_BRACKET = "["
R_BRACKET = "]"
L_PAREN = "("
R_PAREN = ")"
SKIP = "@"
MOLSEP = ";"
AMPERSAND = "&"
DOT = "."

DIHEDRAL_FOR_BOND_SYMBOL = {
    s: (360 // N_DIHEDRALS) * (i - N_DIHEDRALS if i > N_DIHEDRALS // 2 else i)
    for i, s in enumerate(DIHEDRALS)
}
BOND_SYMBOL_FOR_DIHEDRAL = {v: k for k, v in DIHEDRAL_FOR_BOND_SYMBOL.items()}
BOND_SYMBOL_FOR_DIHEDRAL[-180] = E

_pbond = f"(?P<bond>{'|'.join(map(escape, sorted(DIHEDRALS, key=len, reverse=True)))})"
_c = f"[{''.join(map(escape, [PLUS,MINUS,RADICAL,EXTRA_PI,BANG,CW,CCW]))}]*"
_patom = (
    f"(?P<atom>{escape(L_BRACKET)}[0-9]*[A-Za-z][A-Za-z0-9]*{_c}{escape(R_BRACKET)}|"
    f"{escape(L_PAREN)}[0-9A-Za-z]+{escape(R_PAREN)}|[A-Za-z]{_c})"
)
_pring = f"(?P<ring>({escape(L_BRACKET)}[0-9]+{escape(R_BRACKET)}|[3-9]){escape(SKIP)}*)"
_psaturate = f"(?P<saturate>{escape(DOT)})"
_pmolsep = f"(?P<molsep>{escape(MOLSEP)})"
_pampersand = f"(?P<ampersand>{escape(AMPERSAND)})"

RegExp = compile(f"({_pbond}?({_patom}|({_pring})))|{_psaturate}|{_pmolsep}|{_pampersand}")

_dihedral_pat = "|".join(map(escape, sorted(DIHEDRALS, key=len, reverse=True)))
_implicit_c_re = compile(f"({_dihedral_pat})(?={_dihedral_pat})")
_dihedral_set = set(DIHEDRALS)


def _insert_implicit_carbon(s: str) -> str:
    return _implicit_c_re.sub(r"\1C", s)


def _remove_implicit_carbon(t: list[str]) -> list[str]:
    result: list[str] = []
    for i, tok in enumerate(t):
        if (
            tok == "C"
            and len(result) > 0
            and result[-1] in _dihedral_set
            and i + 1 < len(t)
            and t[i + 1] in _dihedral_set
        ):
            continue
        result.append(tok)
    return result


def ToTokens(s: str) -> list[str]:
    """Convert AMSR string to a list of tokens

    :param s: AMSR
    :return: list of tokens
    """
    t = []
    for m in RegExp.finditer(_insert_implicit_carbon(s)):
        g = m.groupdict()
        for k in ["bond", "atom", "saturate", "molsep"]:
            if g[k] is not None:
                t.append(g[k])
        if g["ring"] is not None:
            t.append(g["ring"].replace(SKIP, ""))
            t.extend(SKIP * g["ring"].count(SKIP))
    return t

File: rust_amsr/src/tokens.rs
use regex::Regex;
use lazy_static::lazy_static;

pub const DOT: &str = ".";
pub const DIHEDRALS: [&str; 12] = ["^", "^\\", "<\\", "<", "</", "_/", "_", "\\_", "\\>", ">", "/>", "/^"];
pub const Z: &str = DIHEDRALS[0];
pub const E: &str = DIHEDRALS[6];
pub const CW: &str = "'";
pub const CCW: &str = "`";
pub const PLUS: &str = "+";
pub const MINUS: &str = "-";
pub const EXTRA_PI: &str = ":";
pub const BANG: &str = "!";
pub const RADICAL: &str = "*";
pub const L_BRACKET: &str = "[";
pub const R_BRACKET: &str = "]";
pub const L_PAREN: &str = "(";
pub const R_PAREN: &str = ")";
pub const SKIP: &str = "@";
pub const MOLSEP: &str = ";";
pub const AMPERSAND: &str = "&";

lazy_static! {
    static ref DIHEDRAL_FOR_BOND_SYMBOL: std::collections::HashMap<&'static str, i32> = {
        let mut map = std::collections::HashMap::new();
        for (i, &symbol) in DIHEDRALS.iter().enumerate() {
            let angle = if i > 6 { 30 * (i as i32 - 12) } else { 30 * i as i32 };
            map.insert(symbol, angle);
        }
        map.insert(E, -180);
        map
    };

    static ref BOND_SYMBOL_FOR_DIHEDRAL: std::collections::HashMap<i32, &'static str> = {
        let mut map = std::collections::HashMap::new();
        for (k, v) in DIHEDRAL_FOR_BOND_SYMBOL.iter() {
            map.insert(*v, *k);
        }
        map
    };

    static ref AMSR_REGEX: Regex = {
        let dihedrals_escaped = DIHEDRALS.iter()
            .map(|s| regex::escape(s))
            .collect::<Vec<_>>()
            .join("|");
        
        let bond_pattern = format!(r"(?P<bond>{})", dihedrals_escaped);
        let atom_chars = format!(r"[{}{}{}{}{}{}{}]", 
            regex::escape(PLUS), regex::escape(MINUS), regex::escape(RADICAL),
            regex::escape(EXTRA_PI), regex::escape(BANG), regex::escape(CW), regex::escape(CCW));
        
        let atom_pattern = format!(
            r"(?P<atom>{}[0-9]*[A-Za-z]+{}*{}|{}[0-9A-Za-z]+{}|[A-Za-z]{}*)",
            regex::escape(L_BRACKET), atom_chars, regex::escape(R_BRACKET),
            regex::escape(L_PAREN), regex::escape(R_PAREN), atom_chars
        );
        
        let ring_pattern = format!(
            r"(?P<ring>({}[0-9]+{}|[3-9]){}*)",
            regex::escape(L_BRACKET), regex::escape(R_BRACKET), regex::escape(SKIP)
        );
        
        let saturate_pattern = format!(r"(?P<saturate>{})", regex::escape(DOT));
        let molsep_pattern = format!(r"(?P<molsep>{})", regex::escape(MOLSEP));
        let ampersand_pattern = format!(r"(?P<ampersand>{})", regex::escape(AMPERSAND));
        
        let full_pattern = format!(
            r"({}?({}|({})))|{}|{}|{}",
            bond_pattern, atom_pattern, ring_pattern, saturate_pattern, molsep_pattern, ampersand_pattern
        );
        
        Regex::new(&full_pattern).unwrap()
    };
}

#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Bond(String),
    Atom(String),
    Ring(String),
    Saturate,
    MolSep,
    Ampersand,
}

pub fn to_tokens(s: &str) -> Vec<Token> {
    let mut tokens = Vec::new();
    
    for cap in AMSR_REGEX.captures_iter(s) {
        if let Some(bond) = cap.name("bond") {
            tokens.push(Token::Bond(bond.as_str().to_string()));
        }
        
        if let Some(atom) = cap.name("atom") {
            tokens.push(Token::Atom(atom.as_str().to_string()));
        }
        
        if let Some(ring) = cap.name("ring") {
            let ring_str = ring.as_str();
            let ring_number = ring_str.chars()
                .filter(|c| c.is_ascii_digit())
                .collect::<String>();
            tokens.push(Token::Ring(ring_number));
            
            // Add skip tokens
            let skip_count = ring_str.chars().filter(|c| *c == SKIP.chars().next().unwrap()).count();
            for _ in 0..skip_count {
                tokens.push(Token::Atom(SKIP.to_string()));
            }
        }
        
        if cap.name("saturate").is_some() {
            tokens.push(Token::Saturate);
        }
        
        if cap.name("molsep").is_some() {
            tokens.push(Token::MolSep);
        }
        
        if cap.name("ampersand").is_some() {
            tokens.push(Token::Ampersand);
        }
    }
    
    tokens
}

pub fn get_dihedral_angle(symbol: &str) -> Option<i32> {
    DIHEDRAL_FOR_BOND_SYMBOL.get(symbol).copied()
}

pub fn get_bond_symbol(angle: i32) -> Option<&'static str> {
    BOND_SYMBOL_FOR_DIHEDRAL.get(&angle).copied()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_token_parsing() {
        let amsr = "CCO";
        let tokens = to_tokens(amsr);
        assert_eq!(tokens, vec![
            Token::Atom("C".to_string()),
            Token::Atom("C".to_string()),
            Token::Atom("O".to_string()),
        ]);
    }

    #[test]
    fn test_dihedral_angles() {
        assert_eq!(get_dihedral_angle("^"), Some(0));
        assert_eq!(get_dihedral_angle("_"), Some(180));
        assert_eq!(get_dihedral_angle("E"), Some(-180));
    }
} 